velox/vector/ComplexVector.cpp

```diff
@@ -1781,4 +1782,266 @@ void RowVector::appendNulls(vector_size_t numberOfRows) {
+  auto decodedKeys = DecodedVector(*mapKeys);
+  auto mapValues = mapBase->mapValues();
+  auto decodedValues = DecodedVector(*mapValues);
+  auto offsets = mapBase->rawOffsets();
+  auto sizes = mapBase->rawSizes();
+  vector_size_t outputSize = rows.end();
+
+  for (size_t i = 0; i < keyToIndex.size(); ++i) {
+    children.push_back(
+        BaseVector::create(valueType, outputSize, mapBase->pool()));
+    flatChildren.push_back(children.back()->asFlatVector<ValueType>());
+    VELOX_CHECK_NOT_NULL(flatChildren.back());
+  }
+
+  static const ValueType defaultValue = ValueType();
+  std::vector<bool> visited(keyToIndex.size());
+
+  rows.applyToSelected([&](vector_size_t row) INLINE_LAMBDA {
+    if (decodedVector.isNullAt(row)) {
+      for (auto& child : flatChildren) {
+        if (replaceNulls) {
+          child->set(row, defaultValue);
+        } else {
+          child->setNull(row, true);
+        }
+      }
+      return;
+    }
+    auto decodedIndex = decodedVector.index(row);
+    auto offset = offsets[decodedIndex];
+    auto size = sizes[decodedIndex];
+    std::fill(visited.begin(), visited.end(), false);
+    for (auto j = offset; j < offset + size; ++j) {
+      if (decodedKeys.isNullAt(j) || decodedValues.isNullAt(j)) {
+        continue;
+      }
+      auto key = decodedKeys.valueAt<KeyType>(j);
+      auto it = keyToIndex.find(key);
+      if (it == keyToIndex.end()) {
+        continue;
+      }
+      auto index = it->second;
+      auto value = decodedValues.valueAt<ValueType>(j);
+      visited[index] = true;
+      flatChildren[index]->set(row, value);
+    }
+
+    // Fill missing keys in the row with null or default values.
+    for (size_t field = 0; field < visited.size(); ++field) {
+      if (!visited[field]) {
+        if (replaceNulls) {
+          flatChildren[field]->set(row, defaultValue);
+        } else {
+          flatChildren[field]->setNull(row, true);
+        }
+      }
+    }
+  });
+  return children;
+}
+
+template <TypeKind KeyKind>
+VectorPtr toRowVectorImpl(
+    const BaseVector& vector,
+    const std::vector<variant>& keysToProject,
+    const std::vector<std::string> outputFieldNames,
+    bool replaceNulls,
+    const SelectivityVector& rows) {
+  using KeyType = typename TypeTraits<KeyKind>::NativeType;
+  folly::F14FastMap<KeyType, size_t> lookup;
+  lookup.reserve(keysToProject.size());
+  for (size_t i = 0; i < keysToProject.size(); ++i) {
+    VELOX_CHECK(
+        !keysToProject[i].isNull(), "Keys to project cannot contain null");
+    const KeyType& key = keysToProject[i].value<KeyKind>();
+    // Key must be unique.
+    VELOX_CHECK_EQ(lookup.count(key), 0, "Duplicate keys not allowed");
+    lookup.insert({key, i});
+  }
+
+  std::vector<VectorPtr> children;
+  auto valueType = vector.type()->asMap().valueType();
+  if (valueType->isPrimitiveType() && valueType->kind() != TypeKind::UNKNOWN) {
+    children = VELOX_DYNAMIC_SCALAR_TEMPLATE_TYPE_DISPATCH(
```

Comment:
Is there significant perf benefit in having this separate path? Curious, how much does it save?